src/Storage/StreamWrapper.php

```diff
@@ -0,0 +1,278 @@
+     */
+    public function __destruct()
+    {
+        $this->stream_close();
+    }
+
+    /**
+     * Register a StreamWrapper for reading and writing to Google Storage
+     *
+     * @param string $protocol The name of the protocol to use. Defaults to
+     *        'gs'.
+     * @throws \RuntimeException
+     */
+    public static function register(string $protocol = null)
+    {
+        $protocol = $protocol ?: 'gs';
+        if (!in_array($protocol, stream_get_wrappers())) {
+            if (!stream_wrapper_register($protocol, StreamWrapper::class)) {
+                throw new RuntimeException("Failed to register '$protocol://' protocol");
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Unregisters the SteamWrapper
+     *
+     * @param string $protocol The name of the protocol to unregister. Defaults
+     *        to 'gs'.
+     */
+    public static function unregister(string $protocol = null)
+    {
+        stream_wrapper_unregister($protocol ?: 'gs');
+    }
+
+    /**
+     * Get the default client to use for streams.
+     *
+     * @return StorageClient
+     */
+    public static function getClient()
+    {
+        return self::$client;
+    }
+
+    /**
+     * Set the default client to use for streams.
+     *
+     * @param StorageClient $client
+     */
+    public static function setClient($client)
+    {
+        self::$client = $client;
+    }
+
+    // @codingStandardsIgnoreStart
+    /**
+     * Callback handler for when a stream is opened. For reads, we need to
+     * download the file to see if it can be opened.
+     *
+     * @return bool
+     */
+    public function stream_open($path, $mode, $options, &$opened_path)
+    {
+        // @codingStandardsIgnoreEnd
+        $url = parse_url($path);
+        $this->protocol = $url['scheme'];
+        $this->file = substr($url['path'], 1);
+        $this->mode = $mode;
+
+        $client = $this->getOption('client') ?: self::getClient();
+        $this->bucket = $client->bucket($url['host']);
+
+        if ($this->isWriteable()) {
+            $this->stream = $this->bucket->getStreamableUploader(
+                '',
+                ['name' => $this->file] + $this->getOptions()
+            );
+        } elseif ($this->isReadable()) {
+            try {
+                $this->stream = $this->bucket->object($this->file)->downloadAsStream($this->getOptions());
+            } catch (GoogleException $ex) {
+                return false;
+            }
+        } else {
+            return false;
+        }
+        return true;
+    }
+
+    private function getOption($name, $default = null)
+    {
+        if (array_key_exists($name, $this->getOptions())) {
+            return $this->getOptions()[$name];
+        }
+        return null;
+    }
+
+    private function getOptions()
+    {
+        if (isset($this->options)) {
+            return $this->options;
+        }
+
+        $this->options = [];
```

Comment:
I only want to fetch it once from the context and I'm checking for if we've fetched it by the presence of `$this->options` (4 lines above).